File: org.lttng.scope.ui/src/org/lttng/scope/ui/timeline/widgets/xychart/XYChartFullRangeWidget.java
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.lttng.scope.ui.timeline.widgets.xychart;

import java.util.Objects;
import java.util.stream.Collectors;

import org.eclipse.jdt.annotation.NonNull;
import org.eclipse.jdt.annotation.Nullable;
import org.lttng.scope.ui.timeline.TimelineWidget;

import com.efficios.jabberwocky.common.TimeRange;
import com.efficios.jabberwocky.context.ViewGroupContext;
import com.efficios.jabberwocky.project.TraceProject;
import com.efficios.jabberwocky.views.xychart.control.XYChartControl;
import com.efficios.jabberwocky.views.xychart.model.provider.XYChartModelProvider;
import com.efficios.jabberwocky.views.xychart.model.render.XYChartRender;
import com.efficios.jabberwocky.views.xychart.view.XYChartView;

import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.geometry.Insets;
import javafx.scene.Parent;
import javafx.scene.chart.AreaChart;
import javafx.scene.chart.NumberAxis;
import javafx.scene.chart.XYChart;
import javafx.scene.control.ScrollPane;
import javafx.scene.control.SplitPane;
import javafx.scene.layout.BorderPane;
import javafx.scene.shape.Rectangle;

/**
 * Widget for the timeline view showing data in a XY-Chart. The chart will
 * display data representing the whole trace, and will display highlighted
 * rectangles representing the current visible and selection time ranges.
 */
public final class XYChartFullRangeWidget implements XYChartView, TimelineWidget {

    private static final double CHART_HEIGHT = 50.0;

    private final int weight;

    private final XYChartControl control;
    private final XYChartModelProvider modelProvider;

    private final BorderPane parentPane = new BorderPane();
    private final XYChart<Number, Number> chart;

    private final RedrawTask redrawTask = new RedrawTask();

    public XYChartFullRangeWidget(XYChartControl control, int weight) {
        this.weight = weight;
        this.control = control;
        this.modelProvider = control.getRenderProvider();

        modelProvider.getSeries().get(0).getLineStyle();

        NumberAxis xAxis = new NumberAxis();
        /* Hide the axes in the full range chart */
        xAxis.setTickMarkVisible(false);
        xAxis.setTickLabelsVisible(false);
        xAxis.setOpacity(0);

        NumberAxis yAxis = new NumberAxis();
        xAxis.setAutoRanging(false);
        xAxis.setTickMarkVisible(false);
        xAxis.setTickUnit(0);
        yAxis.setAutoRanging(true);
        yAxis.setTickLabelsVisible(false);
        yAxis.setOpacity(0);

        chart = new AreaChart<>(xAxis, yAxis, null);
//        chart.setTitle(getName());
        chart.setTitle(null);
        chart.setLegendVisible(false);
        chart.setAnimated(false);
        chart.setPadding(new Insets(-10, -10, -10, -10));
//        chart.setStyle("-fx-padding: 1px;"); //$NON-NLS-1$

        chart.setMinHeight(CHART_HEIGHT);
        chart.setPrefHeight(CHART_HEIGHT);
        chart.setMaxHeight(CHART_HEIGHT);

        parentPane.setCenter(chart);
    }

    // ------------------------------------------------------------------------
    // ITimelineWidget
    // ------------------------------------------------------------------------

    @Override
    public String getName() {
        return getControl().getRenderProvider().getProviderName();
    }

    @Override
    public int getWeight() {
        return weight;
    }

    @Override
    public Parent getRootNode() {
        return parentPane;
    }

    @Override
    public void dispose() {
    }

    @Override
    public @NonNull TimelineWidgetUpdateTask getTimelineWidgetUpdateTask() {
        return redrawTask;
    }

    @Override
    public @Nullable SplitPane getSplitPane() {
        /* Not applicable to this widget */
        return null;
    }

    @Override
    public @Nullable ScrollPane getTimeBasedScrollPane() {
        /* Not applicable to this widget */
        return null;
    }

    @Override
    public @Nullable Rectangle getSelectionRectangle() {
        // TODO
        return null;
    }

    @Override
    public @Nullable Rectangle getOngoingSelectionRectangle() {
        // TODO
        return null;
    }

    // ------------------------------------------------------------------------
    // XYChartView
    // ------------------------------------------------------------------------

    @Override
    public XYChartControl getControl() {
        return control;
    }

    @Override
    public ViewGroupContext getViewContext() {
        return getControl().getViewContext();
    }

    @Override
    public void clear() {
//        Platform.runLater(() -> parentPane.getChildren().clear());
    }

    @Override
    public void drawSelection(@NonNull TimeRange arg0) {
        // TODO
    }

    @Override
    public void seekVisibleRange(@NonNull TimeRange timeRange) {

    }

    private class RedrawTask implements TimelineWidget.TimelineWidgetUpdateTask {

        private @Nullable TraceProject<?, ?> lastTraceProject = null;

        @Override
        public void run() {
            TraceProject<?, ?> newTraceProject = getControl().getViewContext().getCurrentTraceProject();
            if (!Objects.equals(newTraceProject, lastTraceProject)) {
                if (newTraceProject != null) {
                    boolean painted = repaintChart(newTraceProject);
                    if (!painted) {
                        newTraceProject = null;
                    }
                }
                lastTraceProject = newTraceProject;
            }
        }

        /**
         * Repaint the whole chart. Should only be necessary if the active trace
         * changes.
         *
         * @return If we have produced a "real" render or not. If not, it might be
         *         because the trace is still being initialized, so we should not
         *         consider it painted.
         */
        private boolean repaintChart(TraceProject<?, ?> traceProject) {
            double viewWidth = parentPane.getWidth();
            TimeRange traceFullRange = TimeRange.of(traceProject.getStartTime(), traceProject.getEndTime());
            long resolution = (long) (traceFullRange.getDuration() / viewWidth) * 1L;
            XYChartRender render = modelProvider.generateRender(modelProvider.getSeries().get(0), traceFullRange, resolution, null);
            if (render == XYChartRender.Companion.getEMPTY_RENDER()) {
                return false;
            }

            ObservableList<XYChart.Data<Number, Number>> data = render.getData().stream()
                    .<XYChart.Data<Number, Number>> map(datapoint -> new XYChart.Data<>(datapoint.getX(), datapoint.getY()))
                    /* Hide the symbols */
                    .peek(xyData -> {
                        Rectangle symbol = new Rectangle(0,0);
                        symbol.setVisible(false);
                        xyData.setNode(symbol);
                    })
                    .collect(Collectors.toCollection(FXCollections::observableArrayList));

            Platform.runLater(() -> {
                chart.setData(FXCollections.observableArrayList());
                XYChart.Series<Number, Number> series = new XYChart.Series<>(data);
                chart.getData().add(series);

                TimeRange range = render.getRange();
                NumberAxis xAxis = (NumberAxis) chart.getXAxis();
                xAxis.setTickUnit(range.getDuration());
                xAxis.setLowerBound(range.getStartTime());
                xAxis.setUpperBound(range.getEndTime());
            });

            return true;
        }

    }

}

File: lttng-scope/src/main/kotlin/org/lttng/scope/views/timeline/widgets/xychart/XYChartFullRangeWidget.kt
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.lttng.scope.views.timeline.widgets.xychart

import com.efficios.jabberwocky.common.TimeRange
import com.efficios.jabberwocky.project.TraceProject
import com.efficios.jabberwocky.views.xychart.control.XYChartControl
import com.efficios.jabberwocky.views.xychart.model.provider.XYChartModelProvider
import com.efficios.jabberwocky.views.xychart.model.render.XYChartRender
import com.efficios.jabberwocky.views.xychart.view.XYChartView
import javafx.application.Platform
import javafx.beans.binding.Bindings
import javafx.collections.FXCollections
import javafx.geometry.Insets
import javafx.scene.chart.AreaChart
import javafx.scene.chart.NumberAxis
import javafx.scene.chart.XYChart
import javafx.scene.layout.BorderPane
import javafx.scene.layout.Pane
import javafx.scene.layout.StackPane
import javafx.scene.paint.Color
import javafx.scene.shape.Rectangle
import javafx.scene.shape.StrokeLineCap
import org.lttng.scope.views.timeline.NavigationAreaWidget
import org.lttng.scope.views.timeline.TimelineWidget

/**
 * Widget for the timeline view showing data in a XY-Chart. The chart will
 * display data representing the whole trace, and will display highlighted
 * rectangles representing the current visible and selection time ranges.
 */
class XYChartFullRangeWidget(override val control: XYChartControl, override val weight: Int) : XYChartView, NavigationAreaWidget {

    companion object {
        private const val CHART_HEIGHT = 50.0

        private val VISIBLE_RANGE_STROKE_WIDTH = 4.0
        private val VISIBLE_RANGE_ARC = 5.0
        private val VISIBLE_RANGE_STROKE_COLOR = Color.GRAY
        private val VISIBLE_RANGE_FILL_COLOR = Color.LIGHTGRAY.deriveColor(0.0, 1.2, 1.0, 0.6)
    }

    private val modelProvider: XYChartModelProvider = control.renderProvider

    override val rootNode = BorderPane()

    private val xAxis = NumberAxis().apply {
        /* Hide the axes in the full range chart */
        isTickMarkVisible = false
        isTickLabelsVisible = false
        tickUnit = 0.0
        opacity = 0.0
        isAutoRanging = false
    }

    private val yAxis = NumberAxis().apply {
        isAutoRanging = true
        isTickLabelsVisible = false
        opacity = 0.0
    }

    private val chart: XYChart<Number, Number> = AreaChart(xAxis, yAxis, null).apply {
//          setTitle(getName())
        title = null
        isLegendVisible = false
        animated = false
        padding = Insets(-10.0, -10.0, -10.0, -10.0)
//          setStyle("-fx-padding: 1px;")

        minHeight = CHART_HEIGHT
        prefHeight = CHART_HEIGHT
        maxHeight = CHART_HEIGHT
    }

    private val visibleRangeRect = Rectangle().apply {
        stroke = VISIBLE_RANGE_STROKE_COLOR
        strokeWidth = VISIBLE_RANGE_STROKE_WIDTH
        strokeLineCap = StrokeLineCap.ROUND
        fill = VISIBLE_RANGE_FILL_COLOR
        arcHeight = VISIBLE_RANGE_ARC
        arcWidth = VISIBLE_RANGE_ARC

        y = 0.0
        heightProperty().bind(Bindings.subtract(chart.heightProperty(), VISIBLE_RANGE_STROKE_WIDTH))

        isVisible = false
        viewContext.currentTraceProjectProperty().addListener { _, _, newVal -> isVisible = (newVal != null) }
    }

    init {
        rootNode.center = StackPane(chart, Pane(visibleRangeRect))
    }

    override fun dispose() {
    }

    // ------------------------------------------------------------------------
    // TimelineWidget
    // ------------------------------------------------------------------------

    override val name = control.renderProvider.providerName

    override val timelineWidgetUpdateTask: TimelineWidget.TimelineWidgetUpdateTask = RedrawTask()

    /** Not applicable to this widget */
    override val splitPane = null

    /* Not applicable to this widget */
    override val timeBasedScrollPane = null

    // TODO
    override val selectionRectangle = null

    // TODO
    override val ongoingSelectionRectangle = null

    // ------------------------------------------------------------------------
    // XYChartView
    // ------------------------------------------------------------------------

    override fun clear() {
        /* Nothing to do, the redraw task will remove all series if the trace is null. */
    }

    override fun drawSelection(selectionRange: TimeRange) {
        // TODO
    }

    override fun seekVisibleRange(newVisibleRange: TimeRange) {
        /* Needs + 10 to be properly aligned, not sure why... */
        val xStart = xAxis.getDisplayPosition(newVisibleRange.startTime) + 10.0
        val xEnd = xAxis.getDisplayPosition(newVisibleRange.endTime) + 10.0
        if (xStart == Double.NaN || xEnd == Double.NaN) return

        with(visibleRangeRect) {
            x = xStart
            width = xEnd - xStart
        }
    }

    private inner class RedrawTask : TimelineWidget.TimelineWidgetUpdateTask {

        private var lastTraceProject: TraceProject<*, *>? = null

        override fun run() {
            var newTraceProject = viewContext.currentTraceProject
            if (newTraceProject == lastTraceProject) return

            if (newTraceProject == null) {
                /* Replace the list of series with an empty list */
                Platform.runLater { chart.data = FXCollections.observableArrayList() }
            } else {
                val painted = repaintChart(newTraceProject)
                if (!painted) {
                    newTraceProject = null
                }
            }
            lastTraceProject = newTraceProject
        }

        /**
         * Repaint the whole chart. Should only be necessary if the active trace
         * changes.
         *
         * @return If we have produced a "real" render or not. If not, it might be
         *         because the trace is still being initialized, so we should not
         *         consider it painted.
         */
        private fun repaintChart(traceProject: TraceProject<*, *>): Boolean {
            val viewWidth = rootNode.width
            val traceFullRange = TimeRange.of(traceProject.startTime, traceProject.endTime)
            val resolution = (traceFullRange.duration / viewWidth).toLong()

            val render = modelProvider.generateRender(modelProvider.series[0], traceFullRange, resolution, null)
            if (render == XYChartRender.EMPTY_RENDER) return false

            val data = render.data
                    .map { XYChart.Data<Number, Number>(it.x, it.y) }
                    /* Hide the symbols */
                    .onEach {
                        val symbol = Rectangle(0.0, 0.0)
                        symbol.isVisible = false
                        it.setNode(symbol)
                    }
                    .toCollection(FXCollections.observableArrayList())

            Platform.runLater {
                chart.data = FXCollections.observableArrayList()
                val series = XYChart.Series(data)
                chart.data.add(series)

                val range = render.range
                with(chart.xAxis as NumberAxis) {
                    tickUnit = range.duration.toDouble()
                    lowerBound = range.startTime.toDouble()
                    upperBound = range.endTime.toDouble()
                }
            }

            return true
        }

    }

}